File: tabuleiro.py
import bispo
import cavalo
import torre


class Tabuleiro:
    def __init__(self):
        # Inicializa a matriz de posição com None (ou qualquer valor inicial desejado)
        self._matrizPosicao = [[0 for _ in range(8)] for _ in range(8)]
        self._pecas = []

    @property
    def matriz_posicao(self):
        return self._matrizPosicao

    @property
    def pecas(self):
        return self._pecas

    @pecas.setter
    def pecas(self, peca):
        indice = len(self._pecas)
        self._pecas.append(peca)
        self._pecas[-1].id = indice
        self._matrizPosicao[peca.posicao_atual_x][peca.posicao_atual_y] = peca

    def validar_movimentacao(self, destino_x, destino_y, peca):
        if self._matrizPosicao[destino_x][destino_y] == 0:
            return True
        else:
            peca_destino = self._matrizPosicao[destino_x][destino_y]
            if peca_destino.cor == peca.cor:
                return False
            else:
                peca_destino.mudar_estado()
                return True

    def movimentar_peca(self, destino_x, destino_y, indice):
        if self._pecas[indice].movimentar(destino_x, destino_y):
            if self.validar_movimentacao(destino_x, destino_y, self._pecas[indice]):
                self._matrizPosicao[destino_x][destino_y] = self._pecas[indice]
                self._matrizPosicao[self._pecas[indice].posicao_origem_x][self._pecas[indice].posicao_origem_y] = 0 # TODO: Verificar porque não apaga a ocorrência no tabuleiro
            return
        # Caso o movimento da peça seja válido mas o tabuleiro na posição de destino não é livre
        print("Movimento Inválido!")
        self._pecas[indice].desfazer_movimento()

    @staticmethod
    def verifica_tipo(elemento):
        if elemento == 0:
            return " \U000025FD "
        else:
            if isinstance(elemento, cavalo.Cavalo):
                if elemento.cor == "Branco":
                    return " \U00002658  "
                else:
                    return " \U0000265E  "
            elif isinstance(elemento, bispo.Bispo):
                if elemento.cor == "Branco":
                    return " \U00002657  "
                else:
                    return " \U0000265D  "
            elif isinstance(elemento, torre.Torre):
                if elemento.cor == "Branco":
                    return " \U00002656  "
                else:
                    return " \U0000265C  "

    def capturados(self, cor):
        for i in range(len(self._pecas)):
            if self._pecas[i].estado is False and self._pecas[i].cor == cor:
                print(f"{self.verifica_tipo(self._pecas[i])}", end="")
        print()

    def imprimir_tabuleiro(self):
        i = 0
        print("  |  1   |  2  |  3  |  4  |  5  |  6  |  7   |  8   |")
        self.capturados("Preto")
        for linha in self._matrizPosicao:
            i += 1
            print(f"{i}", end=" | ")
            for elemento in linha:
                print(f"{self.verifica_tipo(elemento)}", end=" |")
            print()

        self.capturados("Branco")
        print()

File: src/xadrez_jv/Tabuleiro.java
package xadrez_jv;

import java.util.ArrayList;

public class Tabuleiro {
	private Peca[][] matrizPosicao; 
	private ArrayList<Peca> pecas;
	private ArrayList<Peca> capturadasBrancas;
	private ArrayList<Peca> capturadasPretas;
	
	Tabuleiro()
	{
		this.matrizPosicao = new Peca[8][8];
		this.pecas = new ArrayList<>();
		this.capturadasBrancas = new ArrayList<>();
		this.capturadasPretas = new ArrayList<>();
	}
	
	public Peca getMatrizPosicao(int posicaoX, int posicaoY) {
		return matrizPosicao[posicaoX][posicaoY];
	}
	
	public ArrayList<Peca> getPecas() {
		return pecas;
	}
	
	public void setPecas(ArrayList<Peca> pecas) {
		this.pecas = pecas;
	}
	
	public void setPeca(Peca peca)
	{
		int indice = this.pecas.size();
		peca.id = indice;
		this.pecas.add(peca);
		this.matrizPosicao[peca.posicaoAtualX][peca.posicaoAtualY] = peca;
	}
	
	public boolean validarMovimentacao(int destinoX, int destinoY, Peca peca)
	{
		if (this.matrizPosicao[destinoX][destinoY] == null)
			return true;
		
		Peca pecaDestino = this.matrizPosicao[destinoX][destinoY];
		if (pecaDestino.cor == peca.cor)
			return false;
		
		pecaDestino.mudarEstado();
		if (pecaDestino.cor == "Branco") this.capturadasBrancas.add(pecaDestino);
		else this.capturadasPretas.add(pecaDestino);
		return true;
	}
	
	public boolean movimentarPeca(int destinoX, int destinoY, int indice)
	{
		Peca peca = this.pecas.get(indice);
		if (peca.movimentar(destinoX, destinoY))
		{
			if (validarMovimentacao(destinoX, destinoY, peca))
			{
				this.matrizPosicao[destinoX][destinoY] = peca;
				this.matrizPosicao[peca.posicaoOrigemX][peca.posicaoOrigemY] = null;
			}
			return true;
		}
		//Caso o movimento seja válido, mas o tabuleiro na posição de destino não é livre
		this.pecas.get(indice).desfazer_movimento();
		return false;
	}
	
	public boolean isEmpty(int x, int y)
	{
		return (this.matrizPosicao[x][y] == null) ? true : false;
	}
	
	public void limparTabuleiro()
	{
		for (int i = 0; i < 8; i++)
		{
			for (int j = 0; j < 8; j++)
			{
				if (!isEmpty(i, j)) 
					this.matrizPosicao[i][j] = null;
			}
		}
	}
	
	public String verificaTipo(Peca peca)
	{
		if (peca == null)
			return "  ";
		if (peca instanceof Bispo)
		{
			if (peca.cor == "Branco")
				return "BB";
			return "BP";
		}
		if (peca instanceof Cavalo)
		{
			if (peca.cor == "Branco")
				return "CB";
			return "CP";
		}
		if (peca instanceof Torre)
		{
			if (peca.cor == "Branco")
				return "TB";
			return "TP";
		}
		
		return "  ";
	}
	
	public void capturados(String cor)
	{
		Peca peca;
		for (int i = 0; i < this.pecas.size(); i++)
		{
			peca = this.pecas.get(i);
			if (!(peca.estado) && peca.cor == cor)
				System.out.print(verificaTipo(peca));
		}
		System.out.println();
	}
	
	public void imprimirTabuleiro()
	{
		System.out.println("  |  1 | 2 | 3 | 4 | 5 | 6 | 7 | 8 |");
		capturados("Preto");
		for (int i = 0; i < 8; i++)
		{
			System.out.printf("%d | ", i+1);
			for (int j = 0; j < 8; j++)
			{
				System.out.print(verificaTipo(this.matrizPosicao[i][j]) + " |");
			}
			System.out.println();			
		}
		capturados("Branco");
		System.out.println();
	}		
	
	/*public static void main(String args[])
	{
		Tabuleiro t = new Tabuleiro();
		Peca b = new Bispo(0,0,0);
		Peca c = new Cavalo(0, 7, 1);
		Peca to = new Torre(4, 7, 1);
		
		t.setPeca(b); t.setPeca(c); t.setPeca(to);
		t.imprimirTabuleiro();
		
		t.movimentarPeca(7, 7, b.id);
		t.imprimirTabuleiro();
		
		t.movimentarPeca(5, 5, b.id);
		t.imprimirTabuleiro();
		
		t.movimentarPeca(7, 3, b.id);
		t.imprimirTabuleiro();
	}*/
}
